### PR TITLE
Implement fallback to query

### PR DESCRIPTION
**Pull request by pyromaniac**
Sometimes we don't want to sacrifice UX but usage of both (DB and non-DB) uniqueness validators produces 2 DB queries on validation instead of one.
This new `fallback_to_query` option makes DB uniqueness validation to act exactly as the Rails one but with an additional data consistency check layer in case of race conditions.

### CODE REVIEW

**Comment by pyromaniac** [comment]
Btw, `fallback_to_query` might be not the best name for this concept. Maybe something like `use_query_validation`? Or just `use_query`?

**Review by djezzzl — COMMENTED** [review]
I like the overall idea, let's do few things:
1. Rename `fallback_to_query` to `mode:` option with values like: 
- `optimized` - rely only on constraint
- `enhanced` - use default validation + rescue properly from constraint
- `standard` - use default only 

2. Let's update the docs with such changes

WDYT about that? 

And thank you for your contribution!

**Comment by djezzzl** [comment]
BTW, once you push a new commit, the full CI should be started (I updated the configuration).

File: README.md [hunk 1]
```diff
@@ -174,41 +174,45 @@ validates :slug, db_uniqueness: {index_name: :unique_index, case_sensitive: fals
...
+Use `ENV['SKIP_DB_UNIQUENESS_VALIDATOR_INDEX_CHECK'] = 'true'`
 if you want to skip it in some cases. (For example, when you run migrations.)
-- It's two times faster in average because it skips unnecessary SQL query. See benchmarks 
-below for details.
-
-Disadvantages: 
-- Cannot handle multiple database validations at once because database raises 
-only one error per query.
+- It's two times faster in average because it skips unnecessary SQL query. See benchmarks below for details.
+- In enhanced mode if runs the query first, so for UX-critical validations we doin't have to sacrifise anything.
```
Comment (djezzzl) [inline README.md:192]: > if runs 

it runs

> so for UX-critical validations we doin't have to sacrifise anything.

I'd reprase it to: it has different [modes]( # modes) so you can pick up the best for your needs.

File: README.md [hunk 2]
```diff
@@ -174,41 +174,45 @@ validates :slug, db_uniqueness: {index_name: :unique_index, case_sensitive: fals
...
-| allow_blank    | -          | -     | -      | 
+| allow_nil      | -          | -     | -      |
+| allow_blank    | -          | -     | -      |
+
+### Modes
+
+There are 3 `mode` options:
+
+- `:optimized` - the default one. In this mode it turns DB constraint exceptions into proper validation messages.
+- `:enhanced` - a combination of the standard uniqueness validation and the db uniqueness validation. Runs a query first but also rescues from exception. Used best for user-facing validations.
```
Comment (djezzzl) [inline README.md:214]: > Used best for user-facing validations.

I'd rephrase to: The preferable mode for user-facing validations.

File: README.md [hunk 3]
```diff
@@ -174,41 +174,45 @@ validates :slug, db_uniqueness: {index_name: :unique_index, case_sensitive: fals
...
+- `:standard` - in this mode works pretty much the same way as `validates_uniqueness_of` (except the index existence check). Not sure why do we need it though.
```
Comment (djezzzl) [inline README.md:215]: > Not sure why do we need it though.

If we have this in doc, let's better remove this mode at all.

**Review by djezzzl — COMMENTED** [review]
Well done! Just minor comments about documentation and open discussion about `standard` mode. Then I'll just merge it and release.

File: README.md (hunk 3 above)
Comment (pyromaniac) [reply README.md:215]: Well, I spent too much time on it already

**Review by djezzzl — APPROVED** [review]
Well done 👍
